src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java

```diff
@@ -0,0 +1,237 @@
+            spawnDataEntryList.stream()
+                .map(SpawnDataEntry::getGroup)
+                .map(SpawnGroupEntry::getSpawns)
+                .flatMap(List::stream)
+                .filter(spawn -> !blossomConsumed.contains(spawn))
+                .filter(spawn -> BlossomType.valueOf(spawn.getGadgetId()) != null)
+                .forEach(spawn -> {
+                    var type = BlossomType.valueOf(spawn.getGadgetId());
+                    int previewReward = getPreviewReward(type, worldLevel);
+                    blossoms.add(BlossomBriefInfoOuterClass.BlossomBriefInfo.newBuilder()
+                        .setSceneId(sceneId)
+                        .setPos(spawn.getPos().toProto())
+                        .setResin(20)
+                        .setMonsterLevel(monsterLevel)
+                        .setRewardId(previewReward)
+                        .setCircleCampId(type.getCircleCampId())
+                        .setRefreshId(type.getBlossomChestId())  // TODO: replace when using actual leylines
+                        .build()
+                    );
+                });
+        });
+        scene.broadcastPacket(new PacketBlossomBriefInfoNotify(blossoms));
+    }
+
+    public int getWorldLevel(){
+        return scene.getWorld().getWorldLevel();
+    }
+
+    private static Integer getPreviewReward(BlossomType type, int worldLevel) {
+        // TODO: blossoms should be based on their city
+        if (type == null) {
+            Grasscutter.getLogger().error("Illegal blossom type {}",type);
+            return null;
+        }
+
+        int blossomChestId = type.getBlossomChestId();
+        var dataMap = GameData.getBlossomRefreshExcelConfigDataMap();
+        for (var data : dataMap.values()) {
+            if (blossomChestId == data.getBlossomChestId()) {
+                var dropVecList = data.getDropVec();
+                if (worldLevel > dropVecList.length) {
+                    Grasscutter.getLogger().error("Illegal world level {}", worldLevel);
+                    return null;
+                }
+                return dropVecList[worldLevel].getPreviewReward();
+            }
+        }
+        Grasscutter.getLogger().error("Cannot find blossom type {}",type);
+        return null;
+    }
+
+    private static RewardPreviewData getRewardList(BlossomType type, int worldLevel) {
+        Integer previewReward = getPreviewReward(type, worldLevel);
+        if (previewReward == null) return null;
+        return GameData.getRewardPreviewDataMap().get((int) previewReward);
+    }
+
+    public List<GameItem> onReward(Player player, EntityGadget chest, boolean useCondensedResin) {
+        var resinManager = player.getResinManager();
+        synchronized (activeChests) {
+            var it = activeChests.iterator();
+            while (it.hasNext()) {
+                var activeChest = it.next();
+                if (activeChest.getChest() == chest) {
```

Comment:
Not a fan of these iterator patterns, not sure if there's a cleaner way of doing find and delete though.